molecule/src/structure_checker2.cpp

```diff
@@ -0,0 +1,651 @@
+#include "third_party/rapidjson/stringbuffer.h"
+#include "third_party/rapidjson/writer.h"
+#include <functional>
+#include <regex>
+#include <string>
+
+using namespace indigo;
+
+struct CheckParams
+{
+    int check_flags = StructureChecker2::CHECK_ALL;
+    std::vector<int> selected_atoms;
+    std::vector<int> selected_bonds;
+};
+
+static constexpr const char* checkTypeName[] = {
+    StructureChecker2::CHECK_NONE_TXT,      StructureChecker2::CHECK_LOAD_TXT,         StructureChecker2::CHECK_VALENCE_TXT,
+    StructureChecker2::CHECK_RADICAL_TXT,   StructureChecker2::CHECK_PSEUDOATOM_TXT,   StructureChecker2::CHECK_STEREO_TXT,
+    StructureChecker2::CHECK_QUERY_TXT,     StructureChecker2::CHECK_OVERLAP_ATOM_TXT, StructureChecker2::CHECK_OVERLAP_BOND_TXT,
+    StructureChecker2::CHECK_RGROUP_TXT,    StructureChecker2::CHECK_SGROUP_TXT,       StructureChecker2::CHECK_TGROUP_TXT,
+    StructureChecker2::CHECK_CHIRALITY_TXT, StructureChecker2::CHECK_CHIRAL_FLAG_TXT,  StructureChecker2::CHECK_3D_COORD_TXT,
+    StructureChecker2::CHECK_CHARGE_TXT,    StructureChecker2::CHECK_SALT_TXT,         StructureChecker2::CHECK_AMBIGUOUS_H_TXT,
+    StructureChecker2::CHECK_COORD_TXT,     StructureChecker2::CHECK_V3000_TXT,        StructureChecker2::CHECK_ALL_TXT};
+
+static CheckParams check_params_from_string(const char* params)
+{
+    CheckParams r;
+    size_t len = sizeof(checkTypeName) / sizeof(*checkTypeName);
+    if (params)
+    {
+        std::smatch sm1;
+        std::unordered_set<std::string> words;
+        std::string s(params);
+        std::regex rx1(R"(\b(\w+)\b)", std::regex_constants::icase);
+        while (std::regex_search(s, sm1, rx1))
+        {
+            words.insert(sm1[1]);
+            s = sm1.suffix();
+        }
+        r.check_flags = StructureChecker2::CHECK_NONE;
+        for (int i = 0; i < len; i++)
+        {
+            if (words.find(std::string(checkTypeName[i])) != words.end())
+            {
+                r.check_flags |= 1 << i;
+            }
+        }
+        r.check_flags = !r.check_flags || r.check_flags & 1 << (len - 1) ? StructureChecker2::CHECK_ALL
+                                                                         : (r.check_flags == 1 ? StructureChecker2::CHECK_NONE : r.check_flags >> 1);
+
+        std::smatch sm2;
+        s = params;
+        std::regex rx2(R"(\b(atoms|bonds)\b((?:\W+\b\d+\b\W*?)+))", std::regex_constants::icase);
+        std::regex rx3(R"(\b(\d+)\b)");
+        std::smatch sm3;
+        while (std::regex_search(s, sm2, rx2))
+        {
+            std::vector<int>& vec = std::tolower(sm2[1].str()[0]) == 'a' ? r.selected_atoms : r.selected_bonds;
+            std::string a = sm2[2];
+            while (std::regex_search(a, sm3, rx3))
+            {
+                vec.push_back(atoi(sm3[1].str().c_str()));
+                a = sm3.suffix();
+            }
+            s = sm2.suffix();
+        }
+    }
+    return r;
+}
```

Comment:
This logic is pretty complicated, and i think that for first iteration we should not parse this params, and `check2` should just perform all checks for all atoms and bonds.
Atoms and bonds selection will be later done via `selected` field in BaseMolecule (WIP in another branch).